File: src/main/java/ru/citeck/idea/files/scopes/SearchScopeProvider.java
package ru.citeck.idea.files.scopes;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.stream.Collectors;

public class SearchScopeProvider implements com.intellij.psi.search.SearchScopeProvider {

    @Override
    public @Nullable String getDisplayName() {
        return "Citeck";
    }

    @Override
    public @NotNull List<com.intellij.psi.search.SearchScope> getSearchScopes(@NotNull Project project, @NotNull DataContext dataContext) {
        return EcosSearchScope.EP_NAME
            .getExtensionsIfPointIsRegistered()
            .stream()
            .map(eScope -> new SearchScope(eScope, project))
            .collect(Collectors.toList());
    }

}

File: src/main/kotlin/ru/citeck/idea/search/scope/SearchScopeProvider.kt
package ru.citeck.idea.search.scope

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.search.SearchScopeProvider
import icons.Icons
import org.jetbrains.annotations.Nls
import ru.citeck.idea.artifacts.ArtifactTypes
import ru.citeck.idea.artifacts.ArtifactsService
import ru.citeck.idea.project.CiteckProject
import javax.swing.Icon

class SearchScopeProvider : SearchScopeProvider {

    override fun getDisplayName(): String {
        return "Citeck"
    }

    override fun getSearchScopes(
        project: Project,
        dataContext: DataContext
    ): List<com.intellij.psi.search.SearchScope> {
        return listOf(
            CiteckSearchScope(project, "Artifacts"),
            CiteckSearchScope(project, "Data Types", setOf(ArtifactTypes.TYPE_TYPE)),
            CiteckSearchScope(project, "Journals", setOf(ArtifactTypes.TYPE_JOURNAL)),
            CiteckSearchScope(project, "Forms", setOf(ArtifactTypes.TYPE_FORM))
        )
    }

    private class CiteckSearchScope(
        private val project: Project,
        @Nls
        private val name: String,
        private val artifactTypes: Set<String> = emptySet()
    ) : GlobalSearchScope(project) {

        override fun isSearchInModuleContent(aModule: Module): Boolean {
            return CiteckProject.getInstance(project).isCiteckModule(aModule)
        }

        override fun isSearchInLibraries(): Boolean {
            return false
        }

        override fun contains(file: VirtualFile): Boolean {
            val meta = ArtifactsService.getInstance().getArtifactTypeMeta(
                PsiManager.getInstance(project).findFile(file)
            ) ?: return false
            return artifactTypes.isEmpty() || artifactTypes.contains(meta.typeId)
        }

        @Nls
        override fun getDisplayName(): String {
            return name
        }

        override fun getIcon(): Icon {
            return Icons.CiteckLogo
        }
    }
}
